### ISSUE TITLE
Store rank filter with layout

### ISSUE THREAD

**Issue by jlaw9** [Murali-group/GraphSpace#257]
If we could store the rank filter with the layout to be able to share and present smaller sections of the graph (without showing the whole graph or having to upload a new graph with just those paths).

**Comment by adbharadwaj** [comment]
The latest version allows users to store layouts on filtered graphs. I have tried that before.

**Comment by jlaw9** [comment]
Thanks Ted, this is mostly fixed, but doesn't work completely as expected. 

If I set the filter to 100 and then save the layout as _layout1_, the filter is still applied if I reload the page. However, if I change the filter value to 200 and then re-select _layout1_, the filter remains at 200 and is not set back to 100. 

This is not a big issue, but it seems like it could be improved.

**Comment by adbharadwaj** [comment]
@jlaw9 Thanks for testing out this scenario thoroughly. My personal opinion is that layout has nothing to do with filter functionality. When the user filters a graph and stores the layout. They are choosing to store the layout (style and position) not the filter value. 

On the other hand, if we do not change the filter value with the layout, user might get confused. For example, they may see the maximum available value as 200, but not see any node with value k=200. This may confuse them.

**Comment by mitchwagner** [comment]
@adbharadwaj Could you elaborate on your concern about user confusion? Where would they see the maximum rank, other than the slider for rank, such that they would not be able to see a node with that rank?

Or, are you worried that when creating a layout, a user might not be able to see all the nodes if they don't have the rank selector at its highest value? This might be a problem, and perhaps we should force this when a user lays a graph out.

Otherwise, I tend to agree: a graph's layout should be static throughout different ranks, which makes comparing the positions easier. If you want a different layout as you move through ranks, that would be much trickier on a logistical level. It would mean creating an interface to support this, and more storage per graph, especially with a naive implementation.

It seems like what you really want might be some sort of URL option, like ?r=n, where n is the rank, so that you can share a visualization of a graph at a specific rank.

**Comment by jlaw9** [comment]
@mitchwagner, I don't think we want a different layout as you move through the ranks either. The main use case I'm thinking of for this issue is the following: 

1. Set the filter to view the top 20 paths (out of 200)
2. Layout the top 20 paths
3. Save set as default layout and share
4. Instead of seeing the nicely laid out 20 paths when opening the graph, the user or whomever the graph is shared with is greeted with a big mess.

**Comment by adbharadwaj** [comment]
@jlaw9 @mitchwagner  I am not sure how I would implement this feature.

**Comment by jlaw9** [comment]
@adbharadwaj I was thinking of just storing the current filter value in the style/layout JSON when the layout is saved, and then when that layout is opened for the graph, automatically apply the filter level stored in the style JSON. Would that work?

**Comment by adbharadwaj** [comment]
The problem with that solution is that filter is not a layout concept. If we add rank value to layout json, it will violate the fact that currently layout JSON stores only things that can be edited within the layout editor. One possible solution would be to add rank filter widget to layout tool and communicate the consequences properly in the documentation. 

Another issue is that, if we fiddle around with layout_json, it will stop being compatible on Cytoscape desktop and Cytoscape.js. So if a user were to download and try the file somewhere else, he/she wouldnt be able to see the filtered layout.

@tmmurali any thoughts?

**Comment by jlaw9** [comment]
Makes sense to me to have the filter be edited within the layout editor if we're going to allow the user to save the filter.

The rank filter is already not compatible elsewhere. And if we add a field like `"rank_filter": 20` to the top level of the style JSON, I'm pretty sure it will simply be ignored by other Cytoscape.js applications. At least GraphSpace ignores it if I add it to a style JSON file before uploading it. I don't think there would be a problem with it not being compatible.

**Comment by mitchwagner** [comment]
@jlaw9, I am not suggesting a different layout as one moves through the ranks. I am suggesting something like a simple URL option, such that when that specific URL is accessed, the proper number of ranked items is shown according to the layout. Functionally, it seems very similar to what you are suggesting.

1. Set the filter to view the top 20 paths (out of 200)
2. Layout the top 20 paths
3. Save set as default layout and share http:<graph>?k=20
4. The default layout is applied, and the graph is automatically shown with k=20

**Comment by jlaw9** [comment]
Thanks @mitchwagner yeah I agree adding the filter to the url only fixes half the problem. I think a layout database entry that stores the rank filter would work.

So I just tested this issue again by saving a layout with a filter set, and the functionality was better than I thought. If I reload the page with that layout, the filter I saved the layout with is set. What I didn't know is that if another user or a public user opens the graph with that layout selected, the filter is set correctly. I also tested opening the graph without a layout, and the filter is still set correctly after selecting the layout with the filter. 

It seems the only time there is a problem is if the user manually changes the filter and then selects a layout, the filter remains where the user set it to instead of what was saved with the layout - which I think is actually expected. I think this issue can be closed.
